File: SmartWeather/app/src/main/java/com/smart/weather/module/write/WriteFragment.java
package com.smart.weather.module.write;


import android.content.Intent;
import android.content.pm.ActivityInfo;
import android.net.Uri;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.smart.weather.R;
import com.smart.weather.app.MyApp;
import com.smart.weather.base.BaseFragment;
import com.smart.weather.module.write.Views.ToolBean;
import com.smart.weather.module.write.Views.ToolView;
import com.smart.weather.module.write.adapter.WriteAdapter;
import com.smart.weather.module.write.bean.JournalBean;
import com.zhihu.matisse.Matisse;
import com.zhihu.matisse.MimeType;
import com.zhihu.matisse.engine.impl.PicassoEngine;
import com.zhihu.matisse.internal.entity.CaptureStrategy;

import org.apache.commons.collections4.CollectionUtils;
import org.apache.commons.collections4.Predicate;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;

import static android.app.Activity.RESULT_OK;

/**
 * A simple {@link Fragment} subclass.
 * Use the {@link WriteFragment#newInstance} factory method to
 * create an instance of this fragment.
 */
public class WriteFragment extends BaseFragment {
    public static final int REQUEST_CODE_CHOOSE = 9;
    // TODO: Rename parameter arguments, choose names that match
    // the fragment initialization parameters, e.g. ARG_ITEM_NUMBER
    private static final String ARG_PARAM1 = "param1";
    private static final String ARG_PARAM2 = "param2";
    @BindView(R.id.recycleView)
    RecyclerView recycleView;
    Unbinder unbinder;
    @BindView(R.id.toolView)
    ToolView toolView;

    // TODO: Rename and change types of parameters
    private String mParam1;
    private String mParam2;

    private List<JournalBean> writeSectionBeans = new ArrayList<>();
    private WriteAdapter adapter;

    public WriteFragment() {
        // Required empty public constructor
    }

    /**
     * Use this factory method to create a new instance of
     * this fragment using the provided parameters.
     *
     * @return A new instance of fragment WriteFragment.
     */
    // TODO: Rename and change types and number of parameters
    public static WriteFragment newInstance() {
        WriteFragment fragment = new WriteFragment();
        Bundle args = new Bundle();
        args.putString(ARG_PARAM1, "");
        args.putString(ARG_PARAM2, "");
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (getArguments() != null) {
            mParam1 = getArguments().getString(ARG_PARAM1);
            mParam2 = getArguments().getString(ARG_PARAM2);
        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View view = inflater.inflate(R.layout.fragment_write, container, false);
        unbinder = ButterKnife.bind(this, view);
        initView();
        initData();
        return view;
    }

    @Override
    protected void initView() {
        writeSectionBeans.add(new JournalBean(""));

        adapter = new WriteAdapter(writeSectionBeans);
        recycleView.setLayoutManager(new LinearLayoutManager(context));
        recycleView.setAdapter(adapter);

        adapter.notifyDataSetChanged();

        toolView.setDelegate(new ToolView.ToolViewDelegate() {
            @Override
            public void onItemClick(ToolBean toolBean) {

                switch (toolBean.getItemType()){
                    case ToolBean.TOOL_IMAGE:
                        Matisse.from(getActivity())
                                .choose(MimeType.allOf())
                                .countable(true)
                                .maxSelectable(9)
                                .restrictOrientation(ActivityInfo.SCREEN_ORIENTATION_UNSPECIFIED)
                                .thumbnailScale(0.85f)
                                .theme(R.style.Matisse_Zhihu)
                                .imageEngine(new PicassoEngine())
                                .capture(true)
                                .captureStrategy(
                                        new CaptureStrategy(true, MyApp.UPDATE_APP_ID))
                                .forResult(REQUEST_CODE_CHOOSE);
                        break;
                    case ToolBean.TOOL_WEATHER:
                        break;
                }
            }
        });

    }

    @Override
    protected void initData() {

    }

    private void filterJournalItem(){

         Predicate<JournalBean> predicate = new Predicate<JournalBean>() {

                         @Override
                         public boolean evaluate(JournalBean bean) {
                             // TODO Auto-generated method stub
                             return TextUtils.isEmpty(bean.getContent())&&TextUtils.isEmpty(bean.getImageBase64());
                         }
                     };
        List<JournalBean> result = (List<JournalBean>) CollectionUtils.select(writeSectionBeans, predicate);
        writeSectionBeans.removeAll(result);
        writeSectionBeans.add(new JournalBean(""));
        adapter.notifyDataSetChanged();
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == REQUEST_CODE_CHOOSE && resultCode == RESULT_OK) {
            List<Uri> mSelected= Matisse.obtainResult(data);
            for (Uri uri:
                 mSelected) {
                writeSectionBeans.add(new JournalBean("",uri.toString()));
            }
            filterJournalItem();
        }
    }


    @Override
    public void onDestroyView() {
        super.onDestroyView();
        unbinder.unbind();
    }
}

File: MyJournal/app/src/main/java/com/smart/journal/module/write/WriteFragment.kt
package com.smart.journal.module.write


import android.app.Activity.RESULT_OK
import android.content.Intent
import android.os.Bundle
import android.text.TextUtils
import android.view.*
import androidx.annotation.MainThread
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.Observer
import com.alibaba.fastjson.JSON
import com.blankj.utilcode.util.KeyboardUtils
import com.chad.library.adapter.base.BaseQuickAdapter
import com.chad.library.adapter.base.listener.OnItemClickListener
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import com.smart.journal.R
import com.smart.journal.base.BaseFragment
import com.smart.journal.customview.preview.PhotoViewTools
import com.smart.journal.module.map.activity.AMapAdressSearchActivity
import com.smart.journal.module.map.bean.MjPoiItem
import com.smart.journal.module.write.adapter.WriteAdapter
import com.smart.journal.module.write.bean.JournalBean
import com.smart.journal.module.write.bean.ToolBean
import com.smart.journal.module.write.bean.WriteSettingBean
import com.smart.journal.module.write.db.JournalDBHelper
import com.smart.journal.module.write.viewmodel.WriteFragmentViewModel
import com.smart.journal.module.write.views.MoreSettingBottomSheetDialogFragment
import com.smart.journal.module.write.views.ToolView
import com.smart.journal.tools.PermissionTools
import com.smart.journal.tools.eventbus.MessageEvent
import com.smart.journal.tools.file.MJFileTools
import com.smart.journal.tools.image.ImageLoader
import com.smart.journal.tools.logs.LogTools
import com.yanzhenjie.album.Album
import com.yanzhenjie.album.AlbumConfig
import com.yanzhenjie.album.api.widget.Widget
import kotlinx.android.synthetic.main.fragment_write.*
import org.apache.commons.collections4.CollectionUtils
import org.apache.commons.collections4.Predicate
import org.greenrobot.eventbus.EventBus
import java.util.*
import kotlin.collections.ArrayList


/**
 * A simple [Fragment] subclass.
 * Use the [WriteFragment.newInstance] factory method to
 * create an instance of this fragment.
 */
class WriteFragment : BaseFragment() {

    private var mParam1: String? = null
    private var mParam2: String? = null

    private var writeSectionBeans = ArrayList<JournalBean>()
    private var adapter: WriteAdapter? = null
    private var writeSetting: WriteSettingBean? = WriteSettingBean()

    private var isEditable: Boolean = false
        set(value) {
            field = value
            writeSetting!!.isEditable = true
            adapter?.isEditable = field

        }

    private var isShowDataModel = false
    override fun getData() {

    }

    val viewModel by viewModels<WriteFragmentViewModel>()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        EventBus.getDefault().register(this)
        if (arguments != null) {
            mParam1 = arguments!!.getString(ARG_PARAM1)
            mParam2 = arguments!!.getString(ARG_PARAM2)
        }

        viewModel.getJounalData().observe(this, Observer<List<JournalBean>> { journal ->
            // update UI
        })

        viewModel.getLocation().observe(this, Observer<MjPoiItem> {

        })

    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        // Inflate the layout for this fragment
        val view = inflater.inflate(R.layout.fragment_write, container, false)
        initSimpleToolbar(view, writeSetting?.time?.let { Date(it).toLocaleString() })

        return view
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        val showData = activity!!.intent.getSerializableExtra(SHOW_DATA)
        if (showData != null) {
            isShowDataModel = true
            writeSectionBeans = showData as ArrayList<JournalBean>
            activity!!.intent.getSerializableExtra(SHOW_DATA_SETTING)?.let { it ->
                writeSetting = it as WriteSettingBean
            }
            fab.setImageResource(R.drawable.ic_journal_edit)
        } else {
            writeSectionBeans.add(JournalBean(""))
            writeSetting!!.time = System.currentTimeMillis()
            isEditable = true

        }

        init()

    }

    override fun initView() {

        var model: WriteAdapter.WriteAdapterModel? = null
        if (isShowDataModel) {
            model = WriteAdapter.WriteAdapterModel.WriteAdapterModel_SHOW
        } else {
            model = WriteAdapter.WriteAdapterModel.WriteAdapterModel_EDIT
        }
        adapter = WriteAdapter(writeSectionBeans, model!!)
        writeRecycleView!!.layoutManager = androidx.recyclerview.widget.LinearLayoutManager(context)
        writeRecycleView!!.adapter = adapter
        adapter!!.isEditable = isEditable
        writeSetting!!.isEditable = isEditable
        adapter!!.notifyDataSetChanged()

        adapter!!.setOnItemClickListener(object :OnItemClickListener{
            override fun onItemClick(adapter: BaseQuickAdapter<*, *>, view: View, position: Int) {
                if (writeSectionBeans[position].itemType == JournalBean.WRITE_TAG_IMAGE) {
                    PhotoViewTools.showPhotos(object : java.util.ArrayList<String>() {
                        init {
                            writeSectionBeans[position].imageURL?.let { add(it) }
                        }
                    }, 0, context)
                }
            }

        })

        fab.setOnClickListener {
            if (isEditable) {
                saveJournal()
                fab.setImageResource(R.drawable.ic_journal_save)
                KeyboardUtils.hideSoftInput(activity)
            } else {
                isEditable = true
                fab.setImageResource(R.drawable.ic_journal_save)
            }

        }

        toolView!!.setDelegate(object : ToolView.ToolViewDelegate {
            override fun onItemClick(toolBean: ToolBean) {
                if (isEditable) {
                    when (toolBean.itemType) {

                        //位置
                        ToolBean.ToolBeanType.TOOL_LOCATION -> {
                            val locationIntent = Intent(context, AMapAdressSearchActivity::class.java)
                            locationIntent.putExtra(AMapAdressSearchActivity.INTENT_TITLE, "位置")
                            startActivity(locationIntent)
                        }

                        //图片
                        ToolBean.ToolBeanType.TOOL_IMAGE -> {
                            PermissionTools.checkPermission(activity, PermissionTools.PermissionType.PERMISSION_TYPE_STORAGE, object : PermissionTools.PermissionCallBack {
                                override fun permissionYES() {


                                    activity?.let { it ->

                                        Album.initialize(
                                                AlbumConfig.newBuilder(it)
                                                        .setAlbumLoader(ImageLoader())
                                                        .build()
                                        )


                                        Album.image(it) // Image selection.
                                                .multipleChoice()
                                                .camera(true)
                                                .columnCount(4)
                                                .widget(Widget.newDarkBuilder(it).title("选择文件")
                                                        .build())
                                                .onResult {
                                                    LogTools.json(JSON.toJSONString(it))
                                                    val mSelected: ArrayList<String> = ArrayList()
                                                    for (album in it) {

                                                        writeSectionBeans.add(JournalBean("", MJFileTools.saveJournalImageFile2Local(album)))
                                                        filterJournalItem()
                                                    }
                                                }
                                                .start()

                                    }


                                }

                                override fun permissionNO() {

                                }
                            })
                        }


                    }
                }

                if (toolBean.itemType == ToolBean.ToolBeanType.TOOL_MORE) {
                    MoreSettingBottomSheetDialogFragment(writeSetting).show(childFragmentManager, "")
                }

            }
        })


    }

    override fun initData() {

    }

    private fun filterJournalItem() {


        var predicate: Predicate<JournalBean> = Predicate { it ->
            TextUtils.isEmpty(it.content) && TextUtils.isEmpty(it.imageURL)
        }
        val result = CollectionUtils.select(writeSectionBeans, predicate) as List<JournalBean>
        writeSectionBeans.removeAll(result)
        writeSectionBeans.add(JournalBean(""))
        adapter!!.notifyDataSetChanged()


    }


    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {

        if (isShowDataModel) {
            inflater.inflate(R.menu.menu_journal_detail, menu)
        }

    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {

            R.id.toolbar_right_action -> {
            }
            R.id.toolbar_edit -> { //编辑
                isEditable = true

            }

            R.id.toolbar_delete -> {//删除

                writeSetting!!.journalId?.let {
                    JournalDBHelper!!.deleteJournalById(it)
                    finish()
                }
            }

        }


        return false
    }

    private fun saveJournal() {
        JournalDBHelper.saveJournal(writeSectionBeans, settingBean = writeSetting)
        finish()
    }

    private fun finish() {
        EventBus.getDefault().post(MessageEvent("", MessageEvent.NOTE_CHANGE))
        activity!!.finish()
    }


    @MainThread
    override fun onMessageEvent(event: MessageEvent?) {
        super.onMessageEvent(event)

        when (event!!.tag) {
            MessageEvent.NOTE_LOCATION_CHANGE -> {
                val poiItem: MjPoiItem = JSON.parseObject(event.message, MjPoiItem::class.java)
                poiItem?.let {
                    writeSetting!!.location = it
                }
            }
            MessageEvent.NOTE_TAG_CHANGE -> {
                var selectedTag: List<String> = Gson().fromJson(event.message, object : TypeToken<List<String>>() {}.type)
                selectedTag?.let { it ->
                    writeSetting!!.tags = it as List<String>
                }
            }
            MessageEvent.NOTE_FAVOURITE_CHANGE->{
                writeSetting!!.isFavourite = event.message.toInt()==1
            }
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (resultCode != RESULT_OK) {
            return
        }

        when (requestCode) {
            REQUEST_LOCATION_CODE -> {

            }
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        EventBus.getDefault().unregister(this)
    }


    companion object {
        const val REQUEST_LOCATION_CODE = 9

        //日记内容
        const val SHOW_DATA = "showData"

        //日记的设置的内容
        const val SHOW_DATA_SETTING = "showDataSetting"

        // TODO: Rename parameter arguments, choose names that match
        // the fragment initialization parameters, e.g. ARG_ITEM_NUMBER
        private val ARG_PARAM1 = "param1"
        private val ARG_PARAM2 = "param2"


        /**
         * Use this factory method to create a new instance of
         * this fragment using the provided parameters.
         *
         * @return A new instance of fragment WriteFragment.
         */
        // TODO: Rename and change types and number of parameters
        fun newInstance(): WriteFragment {
            val fragment = WriteFragment()
            val args = Bundle()
            args.putString(ARG_PARAM1, "")
            args.putString(ARG_PARAM2, "")
            fragment.arguments = args
            return fragment
        }
    }
}// Required empty public constructor
